sdk/main/src/ECDSAPublicKey.cc

```diff
@@ -99,22 +99,79 @@ std::unique_ptr<PublicKey> ECDSAPublicKey::clone() const
 bool ECDSAPublicKey::verifySignature(const std::vector<unsigned char>& signatureBytes,
                                      const std::vector<unsigned char>& signedBytes) const
 {
+  // incoming signatures are in the raw form (r, s), where r and s are each 32 bytes long
+  if (signatureBytes.size() != 64)
+  {
+    return false;
+  }
+
+  // first, convert the incoming signature to DER format, so that it can be verified
+
+  BIGNUM* signatureR = BN_bin2bn(&signatureBytes.front(), 32, nullptr);
+  if (signatureR == nullptr)
+  {
+    throw std::runtime_error(internal::OpenSSLHasher::getOpenSSLErrorMessage("BN_bin2bn"));
+  }
+
+  BIGNUM* signatureS = BN_bin2bn(&signatureBytes.front() + 32, 32, nullptr);
+  if (signatureS == nullptr)
+  {
+    BN_free(signatureR);
+    throw std::runtime_error(internal::OpenSSLHasher::getOpenSSLErrorMessage("BN_bin2bn"));
+  }
+
+  ECDSA_SIG* signatureObject = ECDSA_SIG_new();
+
+  // this set function transfers ownership of the big numbers to the signature object
+  // after this call succeeds, signatureR and signatureS should NOT be freed manually
+  if (ECDSA_SIG_set0(signatureObject, signatureR, signatureS) <= 0)
+  {
+    BN_free(signatureR);
+    BN_free(signatureS);
+    ECDSA_SIG_free(signatureObject);
+    throw std::runtime_error(internal::OpenSSLHasher::getOpenSSLErrorMessage("BN_bin2bn"));
+  }
+
+  // maximum length of DER encoded signature is 72
+  std::vector<unsigned char> derEncodedSignature(72);
+  unsigned char* encodedSignaturePointer = &derEncodedSignature.front();
+
+  // keep track of how long the DER encoding actually is, since we'll need to tell the verification function
+  int actualSignatureLength = i2d_ECDSA_SIG(signatureObject, &encodedSignaturePointer);
+  if (actualSignatureLength <= 0)
+  {
+    ECDSA_SIG_free(signatureObject);
+    throw std::runtime_error(internal::OpenSSLHasher::getOpenSSLErrorMessage("i2d_ECDSA_SIG"));
+  }
+
+  // wash our hands of the signature object, since we've successfully gotten the DER encoding
+  ECDSA_SIG_free(signatureObject);
+
   EVP_MD_CTX* messageDigestContext = EVP_MD_CTX_new();
 
   if (!messageDigestContext)
   {
     throw std::runtime_error(internal::OpenSSLHasher::getOpenSSLErrorMessage("EVP_MD_CTX_new"));
   }
 
-  if (EVP_DigestVerifyInit(messageDigestContext, nullptr, EVP_sha256(), nullptr, mPublicKey) <= 0)
+  OSSL_LIB_CTX* libraryContext = OSSL_LIB_CTX_new();
```

Comment:
I would check `libraryContext` and `messageDigest` for nullity right after creating them (I believe we do this everywhere else, too). I'm not sure if subsequent calls would just fail if they are null, or if you would get a segfault.

Reply:
Added